--- a/bazel/grpc_build_system.bzl
+++ b/bazel/grpc_build_system.bzl
@@ -27,9 +27,6 @@
 Contains macros used throughout the repo.
 """
 
-load("@build_bazel_apple_support//rules:universal_binary.bzl", "universal_binary")
-load("@build_bazel_rules_apple//apple:ios.bzl", "ios_unit_test")
-load("@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl", "ios_test_runner")
 load("@com_google_protobuf//bazel:cc_proto_library.bzl", "cc_proto_library")
 load("@com_google_protobuf//bazel:upb_c_proto_library.bzl", "upb_c_proto_library")
 load("@com_google_protobuf//bazel:upb_proto_reflection_library.bzl", "upb_proto_reflection_library")
@@ -954,3 +951,7 @@
             "@platforms//cpu:aarch64",
         ],
     )
+
+def ios_unit_test(**kwargs): pass
+def ios_test_runner(**kwargs): pass
+def universal_binary(**kwargs): pass
