--- lib/ipmi_main.c	2021-08-12 21:35:47.000000000 -0700
+++ lib/ipmi_main.c	2021-05-21 14:19:51.000000000 -0700
@@ -44,38 +44,38 @@
 #include <ctype.h>
 #include <locale.h>
 
-#include <ipmitool/helper.h>
-#include <ipmitool/log.h>
-#include <ipmitool/ipmi.h>
-#include <ipmitool/ipmi_intf.h>
-#include <ipmitool/ipmi_session.h>
-#include <ipmitool/ipmi_sdr.h>
-#include <ipmitool/ipmi_gendev.h>
-#include <ipmitool/ipmi_sel.h>
-#include <ipmitool/ipmi_fru.h>
-#include <ipmitool/ipmi_sol.h>
-#include <ipmitool/ipmi_isol.h>
-#include <ipmitool/ipmi_lanp.h>
-#include <ipmitool/ipmi_chassis.h>
-#include <ipmitool/ipmi_mc.h>
-#include <ipmitool/ipmi_firewall.h>
-#include <ipmitool/ipmi_sensor.h>
-#include <ipmitool/ipmi_channel.h>
-#include <ipmitool/ipmi_session.h>
-#include <ipmitool/ipmi_event.h>
-#include <ipmitool/ipmi_user.h>
-#include <ipmitool/ipmi_raw.h>
-#include <ipmitool/ipmi_pef.h>
-#include <ipmitool/ipmi_time.h>
-#include <ipmitool/ipmi_oem.h>
-#include <ipmitool/ipmi_ekanalyzer.h>
-#include <ipmitool/ipmi_picmg.h>
-#include <ipmitool/ipmi_kontronoem.h>
-#include <ipmitool/ipmi_vita.h>
-#include <ipmitool/ipmi_quantaoem.h>
+#include "include/ipmitool/helper.h"
+#include "include/ipmitool/log.h"
+#include "include/ipmitool/ipmi.h"
+#include "include/ipmitool/ipmi_intf.h"
+#include "include/ipmitool/ipmi_session.h"
+#include "include/ipmitool/ipmi_sdr.h"
+#include "include/ipmitool/ipmi_gendev.h"
+#include "include/ipmitool/ipmi_sel.h"
+#include "include/ipmitool/ipmi_fru.h"
+#include "include/ipmitool/ipmi_sol.h"
+#include "include/ipmitool/ipmi_isol.h"
+#include "include/ipmitool/ipmi_lanp.h"
+#include "include/ipmitool/ipmi_chassis.h"
+#include "include/ipmitool/ipmi_mc.h"
+#include "include/ipmitool/ipmi_firewall.h"
+#include "include/ipmitool/ipmi_sensor.h"
+#include "include/ipmitool/ipmi_channel.h"
+#include "include/ipmitool/ipmi_session.h"
+#include "include/ipmitool/ipmi_event.h"
+#include "include/ipmitool/ipmi_user.h"
+#include "include/ipmitool/ipmi_raw.h"
+#include "include/ipmitool/ipmi_pef.h"
+#include "include/ipmitool/ipmi_time.h"
+#include "include/ipmitool/ipmi_oem.h"
+#include "include/ipmitool/ipmi_ekanalyzer.h"
+#include "include/ipmitool/ipmi_picmg.h"
+#include "include/ipmitool/ipmi_kontronoem.h"
+#include "include/ipmitool/ipmi_vita.h"
+#include "include/ipmitool/ipmi_quantaoem.h"
 
 #ifdef HAVE_CONFIG_H
-# include <config.h>
+# include "config.h"
 #endif
 
 #ifdef ENABLE_ALL_OPTIONS
@@ -88,6 +88,8 @@
 void
 ipmi_intf_set_max_request_data_size(struct ipmi_intf * intf, uint16_t size);
 
+extern int verbose;
+extern int csv_output;
 extern const struct valstr ipmi_privlvl_vals[];
 extern const struct valstr ipmi_authtype_session_vals[];
 
@@ -844,12 +846,6 @@
 	/* setup log */
 	log_init(progname, 0, verbose);
 
-	/* load the IANA PEN registry */
-	if (ipmi_oem_info_init()) {
-		lprintf(LOG_ERR, "Failed to initialize the OEM info dictionary");
-		goto out_free;
-	}
-
 	/* run OEM setup if found */
 	if (oemtype &&
 	    ipmi_oem_setup(ipmi_main_intf, oemtype) < 0) {
@@ -1069,8 +1065,6 @@
 		devfile = NULL;
 	}
 
-	ipmi_oem_info_free();
-
 	return rc;
 }
 
