--- lib/ipmi_raw.c.old	2021-08-12 21:56:56.000000000 -0700
+++ lib/ipmi_raw.c	2021-01-12 13:56:28.000000000 -0800
@@ -34,12 +34,12 @@
 #include <stdlib.h>
 #include <stdio.h>
 
-#include <ipmitool/ipmi.h>
-#include <ipmitool/log.h>
-#include <ipmitool/helper.h>
-#include <ipmitool/ipmi_intf.h>
-#include <ipmitool/ipmi_raw.h>
-#include <ipmitool/ipmi_strings.h>
+#include "include/ipmitool/ipmi.h"
+#include "include/ipmitool/log.h"
+#include "include/ipmitool/helper.h"
+#include "include/ipmitool/ipmi_intf.h"
+#include "include/ipmitool/ipmi_raw.h"
+#include "include/ipmitool/ipmi_strings.h"
 
 #define IPMI_I2C_MASTER_MAX_SIZE	0x40 /* 64 bytes */
 
@@ -366,7 +366,7 @@
 		if (is_valid_param(argv[i], &val, "data") != 0)
 			return (-1);
 
-		req.msg.data[i-2] = val;
+		data[i - 2] = val;
 		req.msg.data_len++;
 	}
 
