--- lib/ipmi_user.c.old	2021-08-12 22:08:11.000000000 -0700
+++ lib/ipmi_user.c	2021-01-12 13:56:28.000000000 -0800
@@ -39,14 +39,14 @@
 #include <signal.h>
 #include <unistd.h>
 
-#include <ipmitool/helper.h>
-#include <ipmitool/log.h>
-#include <ipmitool/ipmi.h>
-#include <ipmitool/ipmi_intf.h>
-#include <ipmitool/ipmi_user.h>
-#include <ipmitool/ipmi_constants.h>
-#include <ipmitool/ipmi_strings.h>
-#include <ipmitool/bswap.h>
+#include "include/ipmitool/helper.h"
+#include "include/ipmitool/log.h"
+#include "include/ipmitool/ipmi.h"
+#include "include/ipmitool/ipmi_intf.h"
+#include "include/ipmitool/ipmi_user.h"
+#include "include/ipmitool/ipmi_constants.h"
+#include "include/ipmitool/ipmi_strings.h"
+#include "include/ipmitool/bswap.h"
 
 
 extern int verbose;
@@ -343,11 +343,10 @@
 /*
  * ipmi_user_set_username
  */
-static int
-ipmi_user_set_username(
-		       struct ipmi_intf *intf,
-		       uint8_t user_id,
-		       const char *name)
+int
+_ipmi_user_set_username(struct ipmi_intf *intf,
+			uint8_t user_id,
+			const char *name)
 {
 	struct ipmi_rs	     * rsp;
 	struct ipmi_rq	       req;
@@ -723,7 +722,7 @@
 		return (-1);
 	}
 
-	return ipmi_user_set_username(intf, user_id, argv[3]);
+	return _ipmi_user_set_username(intf, user_id, argv[3]);
 }
 
 /*
