--- src/ipmitool.c	2021-08-12 16:04:42.000000000 -0700
+++ src/ipmitool.c	2020-06-22 15:22:56.000000000 -0700
@@ -33,43 +33,43 @@
  * EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
  */
 
-#include <ipmitool/ipmi.h>
-#include <ipmitool/ipmi_intf.h>
-#include <ipmitool/ipmi_main.h>
+#include "include/ipmitool/ipmi.h"
+#include "include/ipmitool/ipmi_intf.h"
+#include "include/ipmitool/ipmi_main.h"
 
-#include <ipmitool/ipmi_sdr.h>
-#include <ipmitool/ipmi_gendev.h>
-#include <ipmitool/ipmi_sel.h>
-#include <ipmitool/ipmi_fru.h>
-#include <ipmitool/ipmi_sol.h>
-#include <ipmitool/ipmi_isol.h>
-#include <ipmitool/ipmi_tsol.h>
-#include <ipmitool/ipmi_lanp.h>
-#include <ipmitool/ipmi_chassis.h>
-#include <ipmitool/ipmi_mc.h>
-#include <ipmitool/ipmi_sensor.h>
-#include <ipmitool/ipmi_channel.h>
-#include <ipmitool/ipmi_session.h>
-#include <ipmitool/ipmi_event.h>
-#include <ipmitool/ipmi_user.h>
-#include <ipmitool/ipmi_raw.h>
-#include <ipmitool/ipmi_pef.h>
-#include <ipmitool/ipmi_oem.h>
-#include <ipmitool/ipmi_sunoem.h>
-#include <ipmitool/ipmi_fwum.h>
-#include <ipmitool/ipmi_picmg.h>
-#include <ipmitool/ipmi_kontronoem.h>
-#include <ipmitool/ipmi_firewall.h>
-#include <ipmitool/ipmi_hpmfwupg.h>
-#include <ipmitool/ipmi_delloem.h>
-#include <ipmitool/ipmi_ekanalyzer.h>
-#include <ipmitool/ipmi_ime.h>
-#include <ipmitool/ipmi_dcmi.h>
-#include <ipmitool/ipmi_vita.h>
-#include <ipmitool/ipmi_quantaoem.h>
+#include "include/ipmitool/ipmi_sdr.h"
+#include "include/ipmitool/ipmi_gendev.h"
+#include "include/ipmitool/ipmi_sel.h"
+#include "include/ipmitool/ipmi_fru.h"
+#include "include/ipmitool/ipmi_sol.h"
+#include "include/ipmitool/ipmi_isol.h"
+#include "include/ipmitool/ipmi_tsol.h"
+#include "include/ipmitool/ipmi_lanp.h"
+#include "include/ipmitool/ipmi_chassis.h"
+#include "include/ipmitool/ipmi_mc.h"
+#include "include/ipmitool/ipmi_sensor.h"
+#include "include/ipmitool/ipmi_channel.h"
+#include "include/ipmitool/ipmi_session.h"
+#include "include/ipmitool/ipmi_event.h"
+#include "include/ipmitool/ipmi_user.h"
+#include "include/ipmitool/ipmi_raw.h"
+#include "include/ipmitool/ipmi_pef.h"
+#include "include/ipmitool/ipmi_oem.h"
+#include "include/ipmitool/ipmi_sunoem.h"
+#include "include/ipmitool/ipmi_fwum.h"
+#include "include/ipmitool/ipmi_picmg.h"
+#include "include/ipmitool/ipmi_kontronoem.h"
+#include "include/ipmitool/ipmi_firewall.h"
+#include "include/ipmitool/ipmi_hpmfwupg.h"
+#include "include/ipmitool/ipmi_delloem.h"
+#include "include/ipmitool/ipmi_ekanalyzer.h"
+#include "include/ipmitool/ipmi_ime.h"
+#include "include/ipmitool/ipmi_dcmi.h"
+#include "include/ipmitool/ipmi_vita.h"
+#include "include/ipmitool/ipmi_quantaoem.h"
 
 #ifdef HAVE_CONFIG_H
-# include <config.h>
+# include "config.h"
 #endif
 
 #ifdef HAVE_READLINE
