diff --git a/redfishMockupServer.py b/redfishMockupServer.py
index 203fbad..4d56f83 100644
--- a/redfishMockupServer.py
+++ b/redfishMockupServer.py
@@ -31,9 +31,9 @@ from urllib.parse import urlparse, urlunparse, parse_qs
 from redfishMockupServer.rfSsdpServer import RfSSDPServer

 logger = logging.getLogger(__name__)
-logger.setLevel(logging.DEBUG)
+logger.setLevel(logging.CRITICAL)
 ch = logging.StreamHandler(sys.stdout)
-ch.setLevel(logging.INFO)
+ch.setLevel(logging.CRITICAL)
 logger.addHandler(ch)

 tool_version = "1.1.5"
@@ -784,14 +784,14 @@ def main():
                 t2 = threading.Thread(target=mySSDP.start)
                 t2.daemon = True
                 t2.start()
-            logger.info('running Server...')
+            logger.critical('running Server...')
             myServer.serve_forever()

         except KeyboardInterrupt:
             pass

         myServer.server_close()
-        logger.info("Shutting down http server")
+        logger.critical("Shutting down http server")


 # the below is only executed if the program is run as a script
--
2.31.1.498.g6c1eba8ee3d-goog
